File: submit.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>

namespace util {
    namespace parsing {
        template<typename T>
        std::vector<T> parse_line(std::istream& input) {
            std::string line;
            std::getline(input, line);
            std::istringstream iss(line);
            std::vector<T> result;
            T value;
            while (iss >> value) {
                result.push_back(value);
            }
            return result;
        }
    }

    namespace printing {
        template<typename T>
        void print_grid(const std::vector<std::vector<T>>& grid) {
            for (const auto& row : grid) {
                std::cout << "[";
                for (size_t i = 0; i < row.size(); ++i) {
                    std::cout << row[i];
                    if (i < row.size() - 1) std::cout << ", ";
                }
                std::cout << "]" << std::endl;
            }
        }
    }
}

namespace outofplace {
    void solve(std::istream& input, std::ostream& output) {
        std::string line;
        std::getline(input, line); // Read and discard first line

        std::vector<size_t> heights;
        while (std::getline(input, line)) {
            heights.push_back(std::stoull(line));
        }

        int nswaps = 0;
        for (int i = heights.size() - 1; i >= 0; --i) {
            // Find max element in range [0, i)
            int swap_index = -1;
            size_t max_val = 0;
            for (int j = i - 1; j >= 0; --j) {
                if (heights[j] >= max_val) {
                    max_val = heights[j];
                    swap_index = j;
                }
            }

            // Only swap if we found an element greater than heights[i]
            if (swap_index != -1 && heights[i] < heights[swap_index]) {
                std::swap(heights[i], heights[swap_index]);
                nswaps++;
            }
        }

        output << nswaps << std::endl;
    }

    void run_problem() {
        const std::string input_source = "outofplace.in";
        const std::string output_source = "outofplace.out";

        std::istream* input = &std::cin;
        std::ifstream input_file;
        if (input_source != "stdin") {
            input_file.open(input_source);
            input = &input_file;
        }

        std::ostream* output = &std::cout;
        std::ofstream output_file;
        if (output_source != "stdout") {
            output_file.open(output_source);
            output = &output_file;
        }

        solve(*input, *output);
    }
}

int main() {
    outofplace::run_problem();
    return 0;
}

File: submit.rs
pub mod util {
    pub mod parsing {
        pub fn parse_line<T: std::str::FromStr + Default>(
            lines: &mut impl Iterator<Item = std::io::Result<String>>,
        ) -> std::io::Result<Vec<T>> {
            Ok(
                lines
                    .next()
                    .unwrap()?
                    .split_whitespace()
                    .map(|s| s.parse().unwrap_or_default())
                    .collect::<Vec<T>>(),
            )
        }
    }
    pub mod printing {
        pub fn print_grid<T: std::fmt::Debug>(grid: &Vec<Vec<T>>) {
            grid.iter()
                .for_each(|row| {
                    ::std::io::_print(format_args!("{0:?}\n", row));
                });
        }
    }
}
pub mod outofplace {
    use std::{
        io::{self, BufRead, Write},
        mem::swap,
    };
    fn solve(input: Box<dyn BufRead>, mut output: Box<dyn Write>) -> io::Result<()> {
        let mut lines = input.lines();
        _ = lines.next();
        let mut heights: Vec<usize> = lines
            .map(|maybe_line| maybe_line.unwrap().parse().unwrap())
            .collect();
        let mut nswaps = 0;
        for i in (0..heights.len()).rev() {
            if let Some(swap_index) = (0..i)
                .rev()
                .max_by_key(|&j| heights[j])
                .filter(|&j| heights[i] < heights[j])
            {
                heights.swap(i, swap_index);
                nswaps += 1;
            }
        }
        _ = output.write_fmt(format_args!("{0}\n", nswaps));
        Ok(())
    }
    pub fn run_problem() {
        use std::{fs::File, io::{self, BufRead, BufReader, Write}};
        let input_source = "outofplace.in";
        let output_source = "outofplace.out";
        let mut reader: Box<dyn BufRead> = Box::new(io::stdin().lock());
        if input_source != "stdin" {
            let f = File::open(input_source).unwrap();
            reader = Box::new(BufReader::new(f));
        }
        let mut writer: Box<dyn Write> = Box::new(io::stdout());
        if output_source != "stdout" {
            let f = File::create(output_source).unwrap();
            writer = Box::new(f);
        }
        solve(reader, writer).unwrap();
    }
}

fn main() {
    outofplace::run_problem();
}

